resources/build/build-utils.sh

```diff
@@ -465,29 +501,32 @@ builder_parse() {
 
     if [[ $key =~ : ]]; then
       IFS=: read -r action target <<< $key
+      target=:$target
     else
       action="$key"
       target=
     fi
 
     _builder_item_in_array "$action" "${_builder_actions[@]}" && has_action=1 || has_action=0
-    _builder_item_in_array ":$target" "${_builder_targets[@]}" && has_target=1 || has_target=0
+    _builder_item_in_array "$target" "${_builder_targets[@]}" && has_target=1 || has_target=0
 
+    # Expand short -o to --option in options lookup
     if [[ ! -z ${_builder_options_short[$key]+x} ]]; then
-      key="${_builder_options_short[$key]}"
+      key=${_builder_options_short[$key]}
     fi
     _builder_item_in_array "$key" "${_builder_options[@]}" && has_option=1 || has_option=0
 
     if (( has_action )) && (( has_target )); then
+      # apply the selected action and selected target
       _builder_chosen_action_targets+=("$key")
     elif (( has_action )); then
+      # apply the selected action to all targets
       for e in "${_builder_targets[@]}"; do
         _builder_chosen_action_targets+=("$action$e")
       done
     elif (( has_target )); then
-      for e in "${_builder_actions[@]}"; do
-        _builder_chosen_action_targets+=("$e:$target")
-      done
+      # apply the default action to the selected target
+      _builder_chosen_action_targets+=("$_builder_default_action$target")
```

Comment:
How would this handle a case like this?

Script use:
```
> ./build.sh clean :engine
```

Script definition:
```bash
builder_describe \
  "Tests the build-utils.sh builder functions. This is merely an example." \
  "clean        Cleans up any build artifacts" \
  "build+       Do some building" \
  "test         Does some test stuff" \
  ":app" \
  ":engine      Thomas, y'know" \
  "--power,-p   Use powerful mode" \
  "--zoom,-z    Use zoom mode"
```

In my mind, such a call says "only run the `clean` action, and only for the `:engine` target."

Alternatively, I intuitively believe I could use:

```
./build.sh clean build :engine
```

to both `clean` _and_ `build` the `:engine` target, and that target alone.

I bring this up because, by my reading of this part of the code and what precedes it, what we actually get for the first example use is:

```
# From clean -
clean:app
clean:engine

# From the loose :engine - 
build:engine
```

which is completely different.  Assuming I'm interpreting the code correctly.

Reply:
So looks like our internal intuition engines are colliding here.

You are right, currently, this: 

```
> ./build.sh clean :engine
```

does generate this, yes:

```
# From clean -
clean:app
clean:engine

# From the loose :engine - 
build:engine
```

I am open to swapping the inference model to what you've suggested. Pros/cons both ways. Current model leaves each parameter interpreted in isolation which is easy to figure out the action:target pairs. Your model is probably a bit more intuitive but perhaps sticks just a little for more complex invocations, e.g. something like `clean:app build:engine test` -- which targets would apply to `test` here; or am I just being difficult? 🤣)

Reply:
> e.g. something like clean:app build:engine test -- which targets would apply to test here; or am I just being difficult? 🤣)

`:project`, as there's no free-standing target specified otherwise.

Something like `clean:app` effectively scopes those commands together, while free-standing actions and free-standing targets have "general" scope.   If that makes sense?

------

I guess that leads to another niche call structure:

```
> ./build.sh clean:engine :app
```

would yield

```
clean:engine
build:app # due to freestanding :app target with no specified, general action
```

Whereas leaving off the `:app` drops the `build:app` part; there'd be no need for a default action because there's no unpaired (or, un-actioned?) target.  When no target is specified in any fashion, well, `:project` exists; if no action is specified either, `:project` is thus unpaired, so the default action gets paired to it.

Though, I admit, this last example set isn't exactly the clearest scenario.

Reply:
So I feel like either model can lead to confusion, and your suggestion entangles the parameters, whereas in my current model, each parameter can be deduced in isolation, which I think makes it easier to figure out.

So given this, I think I will stick with the existing model: 
* `./build.sh clean :engine` is gonna do `clean:*` and `build:engine`.
* `./build.sh clean:engine` is gonna do `clean:engine`.
* `./build.sh clean build :engine` will give us `clean:*` `build:*` (the final `:engine` parameter is effectively irrelevant).
* `./build.sh` will give us `build:*` (default action is `build`, default target is `*`)

Reply:
As long as this is **_explicitly_** documented in ./build_utils.sh.

Reply:
Is this sufficient? Noting that this is made available with `--help` in any build script, which is more visible than as a comment in build-utils.sh. If this doesn't provide what you are looking for, can you make a suggestion?

https://github.com/keymanapp/keyman/blob/a00c9a9a85f2220ec5fcecad692c0105ae2bfbc9/resources/build/build-utils.sh#L634-L636

Reply:
I do not believe that it is.  Remember, that was in place when I asked the questions that led to this discussion.

Reply:
First draft:

```bash
echo "* Specify `action:target` to run a specific action against a specific target."
echo "* If `action` is specified without a `target` suffix, it will be applied to all targets."
echo "* If `:target` is specified without an `action` prefix, `$_builder_default_action:target` will be inferred."
echo "* If no `action`, `:target`, or `action:target` entries are specified, `$_builder_default_action` will run on all targets."
```